priv/static/scripts/app.js

```diff
@@ -1,87 +1,109 @@
+  },
+  current: {
+    label: document.querySelector("#currentSite"),
+    button: document.querySelector("#current"),
+  },
+  right: {
+    label: document.querySelector("#rightSite"),
+    button: document.querySelector("#right"),
+  },
+  down: {
+    label: document.querySelector("#downSite"),
+    button: document.querySelector("#down"),
+  },
+}
 
-  resetInput()
-
-  // Logic
-  function resetInput() {
-    input.select()
-    submit.removeAttribute("disabled")
-    submit.textContent = "GO"
+// Siteroom stuff
+const DIRS = ["left", "right", "up", "down"]
+function oppositeDirection(dir) {
+  switch (dir) {
+    case "left":
+      return "right";
+    case "right":
+      return "left";
+    case "up":
+      return "down";
+    case "down":
+      return "up";
   }
-
-  function makeXImage() {
-    const x = new Image(10, 10)
-    x.src = "images/x.png"
-    return x
+}
+class Site {
+  constructor(name, { prev, dir } = {}) {
+    this.name = name
+    if (dir) {
+      this[dir] = prev
+    }
+    this.setRelatedSites()
   }
-
-  function createSitelist() {
-    siteList.innerHTML = ''
-    unknown_sites = sites.filter(site => known_sites.includes(site) === false)
-    for (const site of unknown_sites) {
-      const item = document.createElement("li")
-      const anchor = document.createElement("a")
-      anchor.setAttribute("href", `http://${site}`)
-      anchor.setAttribute("target", "_blank")
-      anchor.setAttribute("rel", "noopener noreferrer")
-      const text = document.createTextNode(site)
-      anchor.appendChild(text)
-      item.appendChild(anchor)
-      const x = makeXImage();
-      item.appendChild(x)
-      siteList.appendChild(item)
-
-      x.onclick = function () {
-        item.remove()
-        const siteName = text.textContent
-        known_sites.push(siteName)
-        input.value = siteName
-        fetchSites(siteName)
+  setRelatedSites() {
+    fetchSites(this.name).then(sites => {
+      sites = sites.filter(s => Site.addSite(s))
+      for (const dir of DIRS) {
+        this[dir] = this[dir] || { name: sites.pop() }
+        controls[dir].label.textContent = this[dir].name
+        controls[dir].button.hidden = !this[dir].name
+      }
+    })
+  }
+  move(dir) {
+    let nextCurrent = this[dir]
+    if (!(nextCurrent instanceof Site)) {
+      // make nextCurrent is a proper Site object
+      nextCurrent = new Site(nextCurrent.name, { prev: this, dir: oppositeDirection(dir) })
+      // update current site's neighbor reference to hold proper Site object
+      this[dir] = nextCurrent
+    } else {
+      for (const dir of DIRS) {
+        const name = nextCurrent[dir].name
+        controls[dir].label.textContent = name
+        controls[dir].button.hidden = !name
       }
     }
+    controls.current.label.textContent = nextCurrent.name
+    return nextCurrent
   }
-
-  function fetchSites(site) {
-    if (site === '') {
-      return
+  static addSite(name) {
+    if (Site.sites.has(name)) {
```

Comment:
It doesn't strike me as intuitive that an addSite function should return anything at all. Is that just me?

Reply:
Or what the case would be that it would return false

Reply:
I gotcha. Basically the way I have it working now is that checking if a site is unique + adding it to the set is a single atomic operation. It allows the following line to work nicely:
```js
sites = sites.filter(s => Site.addSite(s))
// now `sites` will be a list of sites that are "new to the graph" (eg. so that we're not creating several foo.com nodes```

Reply:
This can be changed if you'd like. The code above would just become something like:
```
sites = sites.filter(s => Site.isUndiscovered(s))
sites.forEach(s => Site.addSite(s))
```

Reply:
I think I would prefer that unless the former way is a common javascript thing that you think is reasonable. I think the second way is more readable and I don't really see a downside.

Reply:
Yeah I'll change it. It's not a JS thing at all